report/4.Solution/SemanticsDescription.tex

```diff
@@ -0,0 +1,762 @@
+Any function returning a value is an expression.
+For simplicity, all functions will be denoted f in the function declaration transitions.
+
+Declarations of functions must remember the expression directly following the return keyword, this is accomplished by binding the function in the procedure environment with the value of this expression.
+This is not the case for procedures, since they do not return a value.
+The rules for procedures and functions are the same, except for the value $v$ being $none$ in procedure rules.
+
+The transitions are shown below.
+
+\texttt{[$Dcl_PAss_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+		\cfrac
+			{env_V \vdash \langle L, env_P[f \vdash (L, x\sim, v, env_V, env_P)] \rangle \rightarrow_{D_P} env_P'}
+			{env_v \vdash \langle T\ f\ =\ ([T\ x]\sim):\ T\ L;\ S,env_P \rangle \rightarrow_{D_P} env_P'}
+	\end{math}
+	
+	\texttt{ where } $env_V, sto \vdash L \rightarrow_L v$
+\end{center}
+
+\texttt{[$Dcl_PNoAss_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+	\cfrac
+	{env_V \vdash \langle L, env_P[f \vdash (env_V, env_P)] \rangle \rightarrow_{D_P} env_P'}
+	{env_v \vdash \langle T\ f;\ S,env_P \rangle \rightarrow_{D_P} env_P'}
+	\end{math}
+\end{center}
+
+\texttt{[$PreDcl_Ass_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+	\cfrac
+	{env_V \vdash \langle L, env_P[f \vdash (L, x\sim, v, env_V, env_P)] \rangle \rightarrow_{D_P} env_P'}
+	{env_v \vdash \langle f\ =\ ([T\ x]\sim):\ T\ L;\ S,env_P \rangle \rightarrow_{D_P} env_P'}
+	\end{math}
+	
+	\texttt{ where } $env_V, sto \vdash L \rightarrow_L v$
+\end{center}
+
+\texttt{[$Dcl_Pempty_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+		{env_V \vdash \langle \epsilon, env_P \rangle \rightarrow_{D_P}\ env_P'}
+	\end{math}
+\end{center}
+
+\subsection{Lambda Expression}
+Lambda expression in HCL can change only the store by binding a value $v$ to the address $adr$ of the lambda itself in the procedure environment.
+With the the given variable environment $env_V$ and procedure environment $env_P$ the execution of the statement $S$ will change the store $sto$ to $sto'$ and the expression $e$ will be evaluated to $v$.
+The transition-system for $L$ is ($Lambda\ X\ Env_V\ X\ Env_P, \rightarrow_L, sto \cup Lit$).
+The transitions must therefore be on the form.
+\begin{center}
+	$env_V, env_P \vdash \langle S, sto \rangle \rightarrow_L sto'$
+\end{center}
+
+The transitions are shown below.
+
+\texttt{[$Lmb-ExpRetn_{BSS}$]}
+\begin{center}
+	\begin{math}
+		\cfrac
+			{env_V, env_P \vdash \langle S, sto \rangle \rightarrow_L sto', v}
+			{env_V, env_P \vdash \langle \{\ S\ return\ e \}, sto \rangle \rightarrow_L sto', v}
+	\end{math}
+	
+	\texttt{ where } $ env_V, sto' \rightarrow_e v$
+\end{center}
+
+\texttt{[$Lmb-ImpRetn_{BSS}$]}
+\begin{center}
+	\begin{math}
+	\cfrac
+	{env_V, env_P \vdash \langle S, sto \rangle \rightarrow_L sto', v}
+	{env_V, env_P \vdash \langle \{\ S\ e\ \}, sto \rangle \rightarrow_L sto', v}
```

Comment:
Implicit return er kun for lambdas med 1 statement. Der kan ikke være statements foran e.